algosdk/abi/method.py

```diff
@@ -0,0 +1,219 @@
+import json
+
+from Cryptodome.Hash import SHA512
+
+from algosdk import abi, constants, error
+
+
+TRANSACTION_ARGS = (
+    "txn",  # Denotes a placeholder for any of the six transaction types below
+    constants.PAYMENT_TXN,
+    constants.KEYREG_TXN,
+    constants.ASSETCONFIG_TXN,
+    constants.ASSETTRANSFER_TXN,
+    constants.ASSETFREEZE_TXN,
+    constants.APPCALL_TXN,
+)
+
+
+class Method:
+    """
+    Represents a ABI method description.
+
+    Args:
+        name (string): name of the method
+        args (list): list of Argument objects with type, name, and optional
+        description
+        returns (Returns): a Returns object with a type and optional description
+        desc (string, optional): optional description of the method
+    """
+
+    def __init__(self, name, args, returns, desc=None) -> None:
+        self.name = name
+        self.args = args
+        self.desc = desc
+        self.returns = returns
+        # Calculate number of method calls passed in as arguments and
+        # add one for this method call itself.
+        txn_count = 1
+        for arg in self.args:
+            if arg.type in TRANSACTION_ARGS:
+                txn_count += 1
+        self.txn_calls = txn_count
+
+    def __eq__(self, o: object) -> bool:
+        if not isinstance(o, Method):
+            return False
+        return (
+            self.name == o.name
+            and self.args == o.args
+            and self.returns == o.returns
+            and self.desc == o.desc
+            and self.txn_calls == o.txn_calls
+        )
+
+    def get_signature(self):
+        arg_string = ",".join([str(arg.type) for arg in self.args])
+        ret_string = self.returns.type if self.returns else "void"
+        return "{}({}){}".format(self.name, arg_string, ret_string)
+
+    def get_selector(self):
+        """
+        Returns the ABI method signature, which is the first four bytes of the
+        SHA-512/256 hash of the method signature.
+
+        Returns:
+            bytes: first four bytes of the method signature hash
+        """
+        hash = SHA512.new(truncate="256")
+        hash.update((self.get_signature()).encode("utf-8"))
+        return hash.digest()[:4]
+
+    def get_txn_calls(self):
+        """
+        Returns the number of transactions needed to invoke this ABI method.
+        """
+        return self.txn_calls
+
+    @staticmethod
+    def _parse_string(s):
+        # Parses a method signature into three tokens, returned as a list:
+        # e.g. 'a(b,c)d' -> ['a', 'b,c', 'd']
```

Comment:
Will we ever get nested parens?

Reply:
There could be nested tuples for arguments such as: https://github.com/algorand/py-algorand-sdk/blob/155fc9002f1b606843de10a50224f817f792327e/test_unit.py#L4207

As an aside, transaction arguments (`pay`, `appl`, etc.) or foreign objects (`apfa` ...) cannot be nested inside a parentheses, and the ABI encoding/typing method will throw an exception if someone tries to do so.

Reply:
I don't think the _arguments_ are nested as these are just `"(string,uint16),bool"`

Reply:
Fair, it could look something more like: https://github.com/algorand/py-algorand-sdk/blob/8cd3bb615ca3c1c15846bd9eb7ec180ac4267fc4/test_unit.py#L3720 

Funnily, we can use `f-strings` in the unit tests because they do not need to be externally exposed 😅